semver.py

```diff
@@ -615,6 +643,44 @@ def finalize_version(version):
     return format_version(verinfo['major'], verinfo['minor'], verinfo['patch'])
 
 
+def replace(version, **parts):
+    """Replace one or more parts of a version and return the new string
+
+    :param str version: the version string to replace
+    :param dict parts: the parts to be updated. Valid keys are:
+      ``major``, ``minor``, ``patch``, ``prerelease``, or ``build``
+    :return: the replaced version string
+    :raises: KeyError, when parts does not contain valid keys
+    :rtype: str
+
+    >>> import semver
+    >>> semver.replace("1.2.3", major=2, patch=10)
+    '2.2.10'
+    """
+    result = {}
```

Comment:
Why isn't the existing implementation reused? `return str(parse(version).replace(**parts))`

I would strongly oppose having 2 implementations of the same thing.

Reply:
> Why isn't the existing implementation reused? `return str(parse(version).replace(**parts))`

Sorry, either I miss something, or this doesn't work.

The expression `parse(version)` returns a `dict` and that doesn't have a member function `.replace`. :wink: You probably mean something like:

```python
parse(version).update(**parts)
```

Unfortunately, even this won't work. The function `.update` updates the `dict` _in place_, returns `None`. What was your idea?

I agree though, we shouldn't have two implementations for the same thing. My idea was to check for invalid keys for the `parts` part. If you have a clever idea, I will be happy to include it.

Reply:
Essentially my idea would be: Create a VersionInfo object, operate on that, convert to output (either a VersionInfo object or a string, as in this case). So: `str(parse_version_info).replace(**parts))`

I don't really understand the design decision that the shortest function `parse` returns a dict rather than a VersionInfo object but I guess this cannot be changed now. I'd always create a proper object first and then do manipulations on that.

This would make bump_version etc. also simpler.

Reply:
I think, I took your idea and reduced it to the bare minimum:

```python
def replace(version, **parts):
    """Replace one or more parts of a version and return the new string
    [...]
    """
    version = parse_version_info(version)
    return str(version.replace(**parts))
```

However, when I try to pass invalid keys, I got a `TypeError`:

```python
>>> semver.replace("1.2.3", major=2, patch=10, answer=42)
Traceback (most recent call last)
...
TypeError: __init__() got an unexpected keyword argument 'foo'
```

Looks a bit ugly, doesn't it? :wink:

Reply:
Yes, this was what I had in mind.

Oh ok, so a wrong keyword is a TypeError not a KeyError... Well if you want it to be a KeyError and/or make the message better you need to adjust the replace function as outlined in https://github.com/k-bx/python-semver/pull/156/files#r332368167

However as Python is raising a TypeError in such cases I'd vote for a typeerror too and make the error message match: `TypeError(replace() got an unexpected keyword argument 'answer')`

Reply:
I used your last suggestion and raised a `TypeError`.